### ISSUE TITLE
WithCustomClaims in Verify token options should accept a type instead of pointer

### ISSUE THREAD

**Issue by vijathanga** [clerk/clerk-sdk-go#202]
Currently `WithCustomClaims` function in `token_option.go` accepts a pointer to a struct variable. So we have to define and pass the struct pointer when initializing the `WithSessionV2` middleware. The same struct pointer is used for unmarshalling the custom claims for all incoming requests. This can cause consistency issues when multiple requests are processed at the same time.

This can be solved by either
 1. Accept a Custom claim struct type in `WithCustomClaims` function and dynamically create, populate the struct and set the value in context
 2. Provide a function hook for dealing with custom claims

**Comment by georgepsarakis** [comment]
Hello @vijathanga ! Sorry that you encountered this issue and thanks for reporting. Your observation is correct:

> This can cause consistency issues when multiple requests are processed at the same time.

The `WithCustomClaims` option was designed for direct calls to `VerifyToken` since the custom claims are assigned directly to the referenced struct instance which would be constructed before each `VerifyToken` call:

https://github.com/clerk/clerk-sdk-go/blob/2f041fef8d673473d84d5b8360b0650d43b11ae3/clerk/clerk.go#L51

> Accept a Custom claim struct type in WithCustomClaims function

Do you perhaps mean a constructor function for new objects that will be populated with the custom claims? Something like:

```go
WithSessionV2(client, WithSessionClaimsFunc(func() interface{} {
  return &MyCustomClaims{}
})
```

> populate the struct and set the value in context

Would exposing the custom claims as a field in the `SessionClaims` struct that's already available in the context work instead?

https://github.com/clerk/clerk-sdk-go/blob/2f041fef8d673473d84d5b8360b0650d43b11ae3/clerk/middleware_v2.go#L38

So to sum up the approach assuming a custom struct `MyCustomClaims`:

```go
client, err := clerk.NewClient(apiKey)
// Configure the middleware
clerk.WithSessionV2(client, WithSessionClaimsFunc(func() interface{} {
  return &MyCustomClaims{}
})

// In the request
sessionClaims, ok := clerk.SessionFromContext(req.Context())
if !ok {
  // unauthenticated
}
customClaims, ok := sessionClaims.Custom.(*MyCustomClaims)
if ok {
  // use customClaims
}
```

Mostly asking if the final snippet would address your use case, this change will require some additional design considerations from our side!

**Comment by vijathanga** [comment]
Hi @georgepsarakis, I appreciate the quick response.

The final snippet looks good, It should satisfy our needs.

**Comment by georgepsarakis** [comment]
Thanks for the quick follow-up @vijathanga , I've opened a ticket internally so that we can look into this.

**Comment by gkats** [comment]
@vijathanga Hello, sorry this took so long. We were preparing a new version for the Go SDK. 

Here's an attempt to solve the concurrent access issue by providing a constructor for the custom claims. 
https://github.com/clerk/clerk-sdk-go/pull/253

Feel free to provide feedback and follow along with the discussion.

**Comment by gkats** [comment]
Closing this since it is fixed in v2.
